Error on invalid alphanumeric token for crates.io

Akida31 opened the pull request:
ref #11571 

When using `cargo login` and calling an api which requires authentification there will be an error if the given token is not a valid alphanumerical string.
This check is only enabled for crates.io because
only for that registry we can be certain, that the generated token should have been alphanumeric, see [the code here](https://github.com/rust-lang/crates.io/blob/7ea41e9d345f05566ee776b7cbb62e46ccf6b393/src/util/token.rs#L15). So if I'm not mistaken, this should not be a breaking change, since crates.io only generates fitting tokens. (Should I add a comment to the crates.io code that modifying this logic can break cargo?)

I'm not sure if the fix works and is enough to close the issue, please say if you have any corrections or improvements!

I don't know if the check should also be enabled for other registries and it would be really bad if the check is too strict.
In the linked issue it was recommended to encode invalid characters, but I don't know in which encoding. I saw in [this http rfc](https://www.rfc-editor.org/rfc/rfc7230#section-3.2.4) that only the ISO-8859-1 charset is allowed and everything else must be [encoded](https://www.rfc-editor.org/rfc/rfc7230#section-3.2.4) but this seems somewhat complex and hard to implement. There is a crate `rust-encoding` which should be capable doing this (from a first look), but I don't know if a new dependency only for this is justified. There seems to be `percent encoding` already in the dependency tree but I have no idea if it would be correct and work.
If you have any idea about this encoding, please say so.

r? @Eh2406 (since you suggested the encoding part)

Eh2406 commented:
Thank you for working on this!

I think this is a good change and a big step in the right direction. But I think we can go further. If the HTTP spec requires ISO-8859-1 charset, then I think cargo should error if you give it a token that is not in that range. (I do not think cargo should do any escaping.) This would technically be a breaking change, but only for registries that use very odd tokens. How hard would it be to check just that property?

If it's doable, I can bring up the breaking change with the Team.

Akida31 commented:
Thanks for your reply and sorry for the late answer, I had not worked with encodings before so I had to read some things (and had many other things to do ;) ).

If I understand ISO 8859-1 correctly, only bytes in the range 0x20-0x7e and 0xa0-0xff are allowed (defined). But our `String` is utf-8 and for that it looks like the last valid code point for chars with a size of 1 byte is 0x7e.

I've created a test here: [commit](https://github.com/Akida31/cargo/commit/179d8d3cfde27271fcc8786c8bc15b0cfb9dcd50) / [files](https://github.com/Akida31/cargo/tree/test_server/TEST)
It looks like flask accepts all header values (even if they are invalid per spec) and axum accepts only the following ([source](https://docs.rs/http/0.2.8/src/http/header/value.rs.html#588-590))

`axum` and `crates.io` use the `http` crate for validation of header values:
```rust
fn is_valid(b: u8) -> bool {
    b >= 32 && b != 127 || b == b'\t'
}
```



The python requests library only permits unicode values up to 0xff.
So I would recommend restricting the token to
```rust
fn check(b: u8) -> bool {
    b >= 32 // undefined in ISO-8859-1, in ASCII/ UTF-8 not-printable character
    && b < 128 // utf-8: the first bit signals a multi-byte character
    && b != 127 // 127 is a control character in ascii and not in ISO 8859-1
    || b == b't' // tab is also allowed (even when < 32)
}
```
As this is more restrictive than the `is_valid` check in the crates.io server more tokens would be denied from cargo.
But crates.io tokens are currently only alphanumeric so this should not be a problem in the future. [alexandrie](https://github.com/Hirevo/alexandrie/blob/d939ba49bb7e65880b0f2539cdc45f3c8a2dacba/crates/alexandrie/src/utils/auth/mod.rs#L14]) (an alternative registry) uses only alphanumeric tokens too.
So it could make sense to restrict tokens to alphanumeric, since it would make the logic easier.

This is my opinion and I have not that much experience with that topic :)
If you think that the code above is the way to proceed, please say so and I will implement it.

Eh2406 commented:
Thank you for doing the research! I agree, the correct approach is for cargo to check the token as inputted for all registries and make sure it passes the `check` function you defined. We could be more restrictive for crates.io, but I don't know if it's worth it.

This is technically a breaking change. But I don't think it will affect users doing anything reasonable so I think it is worth doing. Any registry uses other characters is just asking for HTTP Confucian attacks.

However, that is my opinion. I will nominate for the cargo team to discuss in a meeting.

Eh2406 commented:
The cargo team discussed this proposal today. We agreed to require that all tokens match your proposed `check` function! If this breaks someone they can file an issue and suggest a new standard, but I strongly suspect this will not break anyone.

Akida31 commented:
I've rebased and implemented the check as written.
I'm not quite sure about the error message (in crates/crates_io/lib.rs:531). It is rather unhelpful but I think printing the token wouldn't be that secure.
Also I don't know what the best solution is for src/cargo/ops/registry.rs:904: Exposing the token isn't that nice but Secure is not available in crates/crates_io/lib.rs

Eh2406 commented on src/cargo/ops/registry.rs:
```diff
@@ -901,6 +901,7 @@ pub fn registry_login(
             if tok.is_empty() {
```
Let's move this `is_empty` check into `check_token`.

Eh2406 replied:
This code can be removed now.

Eh2406 commented on crates/crates-io/lib.rs:
```diff
@@ -510,3 +511,23 @@ pub fn is_url_crates_io(url: &str) -> bool {
+    if is_valid {
+        Ok(())
+    } else {
+        Err(anyhow::anyhow!("invalid token."))
```
Perhaps `token contains invalid characters.` or `token must contain only printable ISO-8859-1 characters as it is sent in a HTTPS header.`

Eh2406 commented on tests/testsuite/login.rs:
```diff
@@ -126,6 +126,38 @@ fn empty_login_token() {
+            .run();
+    };
+    let check = |stdin: &str| {
+        check_(stdin, "[ERROR] invalid token.");
```
I could go either way on whether this helper function pulls its weight.
I'm leaning toward just using the two argument version even though it means repeating `"[ERROR] invalid token."`. But certainly could be convinced otherwise.

Eh2406 reviewed (approved):
Thank you so much, this is looking almost there.

Akida31 commented:
* moved `is_empty` check into `check_token`
* improved error message (is quite long now but should explain the error well)
* removed one helper function from new test

If you have any other improvements, please say so!

And I don't know how the normal process is, if this somewhat breaking change should get a short note in release notes or somewhere else.

Eh2406 commented:
@bors r+

bors commented:
:pushpin: Commit 8502fa85689015234c9d5c84e3957ff32f03e072 has been approved by `Eh2406`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=Eh2406 8502fa85689015234c9d5c84e3957ff32f03e072 -->
<!-- homu: {"type":"Approved","sha":"8502fa85689015234c9d5c84e3957ff32f03e072","approver":"Eh2406","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 8502fa85689015234c9d5c84e3957ff32f03e072 with merge a66f1233f9ad61e91cc8fd86f0b6bd15e66443a8...
<!-- homu: {"type":"BuildStarted","head_sha":"8502fa85689015234c9d5c84e3957ff32f03e072","merge_sha":"a66f1233f9ad61e91cc8fd86f0b6bd15e66443a8"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/4197619390/jobs/7280962833)
Approved by: Eh2406
Pushing a66f1233f9ad61e91cc8fd86f0b6bd15e66443a8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Eh2406","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/4197619390/jobs/7280962833"},"merge_sha":"a66f1233f9ad61e91cc8fd86f0b6bd15e66443a8"} -->

Akida31 commented:
Thanks for your review!

weihanglo commented:
I guess we forgot to bump the version of `crates-io` crate, no?
